sentry: conn_executor.go:431: panic while executing 1 statements: UPSERT INTO _(_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _) VALUES ($1, $2, __more30__), (__more1__): caused by <redacted>

This issue was autofiled by Sentry. It represents a crash or reported error on a live cluster with telemetry enabled.

Sentry link: https://sentry.io/organizations/cockroach-labs/issues/1095469822/?referrer=webhooks_plugin

Panic message:

>conn_executor.go:431: panic while executing 1 statements: UPSERT INTO _(_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _) VALUES ($1, $2, __more30__), (__more1__): caused by <redacted>


<details>
<summary>Stacktrace (expand for inline code snippets):</summary>

https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/conn_executor.go#L430-L432 in pkg/sql.(*Server).ServeConn.func1
/usr/local/go/src/runtime/asm_amd64.s#L521-L523 in runtime.call32
/usr/local/go/src/runtime/panic.go#L512-L514 in runtime.gopanic
/usr/local/go/src/runtime/panic.go#L43-L45 in runtime.panicindex
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/row/fetcher.go#L684-L686 in pkg/sql/row.(*Fetcher).processKV
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/row/fetcher.go#L978-L980 in pkg/sql/row.(*Fetcher).NextRow
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/distsqlrun/tablereader.go#L159-L161 in pkg/sql/distsqlrun.(*rowFetcherWrapper).Next
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/distsqlrun/joinreader.go#L412-L414 in pkg/sql/distsqlrun.(*joinReader).performLookup
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/distsqlrun/joinreader.go#L302-L304 in pkg/sql/distsqlrun.(*joinReader).Next
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/row_source_to_plan_node.go#L77-L79 in pkg/sql.(*rowSourceToPlanNode).Next
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/upsert.go#L308-L310 in pkg/sql.(*upsertNode).BatchedNext
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/plan_batch.go#L172-L174 in pkg/sql.(*rowCountNode).startExec
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/plan.go#L495-L497 in pkg/sql.startExec.func2
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/walk.go#L144-L146 in pkg/sql.(*planVisitor).visitInternal.func1
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/walk.go#L615-L617 in pkg/sql.(*planVisitor).visitInternal
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/walk.go#L111-L113 in pkg/sql.(*planVisitor).visit
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/walk.go#L75-L77 in pkg/sql.walkPlan
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/plan.go#L498-L500 in pkg/sql.startExec
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/plan_node_to_row_source.go#L123-L125 in pkg/sql.(*planNodeToRowSource).Start
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/distsqlrun/processors.go#L799-L801 in pkg/sql/distsqlrun.(*ProcessorBase).Run
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/distsqlrun/flow.go#L625-L627 in pkg/sql/distsqlrun.(*Flow).Run
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/distsql_running.go#L251-L253 in pkg/sql.(*DistSQLPlanner).Run
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/distsql_running.go#L838-L840 in pkg/sql.(*DistSQLPlanner).PlanAndRun
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/conn_executor_exec.go#L1111-L1113 in pkg/sql.(*connExecutor).execWithDistSQLEngine
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/conn_executor_exec.go#L947-L949 in pkg/sql.(*connExecutor).dispatchToExecutionEngine
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/conn_executor_exec.go#L455-L457 in pkg/sql.(*connExecutor).execStmtInOpenState
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/conn_executor_exec.go#L101-L103 in pkg/sql.(*connExecutor).execStmt
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/conn_executor.go#L1240-L1242 in pkg/sql.(*connExecutor).run
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/conn_executor.go#L432-L434 in pkg/sql.(*Server).ServeConn
https://github.com/cockroachdb/cockroach/blob/25dd36f0139bf65b80758deeeccf35ee17ebd622/pkg/sql/pgwire/conn.go#L336-L338 in pkg/sql/pgwire.(*conn).serveImpl.func4
</details>

```
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn.func1 at line 431
/usr/local/go/src/runtime/asm_amd64.s in runtime.call32 at line 522
/usr/local/go/src/runtime/panic.go in runtime.gopanic at line 513
/usr/local/go/src/runtime/panic.go in runtime.panicindex at line 44
pkg/sql/row/fetcher.go in pkg/sql/row.(*Fetcher).processKV at line 685
pkg/sql/row/fetcher.go in pkg/sql/row.(*Fetcher).NextRow at line 979
pkg/sql/distsqlrun/tablereader.go in pkg/sql/distsqlrun.(*rowFetcherWrapper).Next at line 160
pkg/sql/distsqlrun/joinreader.go in pkg/sql/distsqlrun.(*joinReader).performLookup at line 413
pkg/sql/distsqlrun/joinreader.go in pkg/sql/distsqlrun.(*joinReader).Next at line 303
pkg/sql/row_source_to_plan_node.go in pkg/sql.(*rowSourceToPlanNode).Next at line 78
pkg/sql/upsert.go in pkg/sql.(*upsertNode).BatchedNext at line 309
pkg/sql/plan_batch.go in pkg/sql.(*rowCountNode).startExec at line 173
pkg/sql/plan.go in pkg/sql.startExec.func2 at line 496
pkg/sql/walk.go in pkg/sql.(*planVisitor).visitInternal.func1 at line 145
pkg/sql/walk.go in pkg/sql.(*planVisitor).visitInternal at line 616
pkg/sql/walk.go in pkg/sql.(*planVisitor).visit at line 112
pkg/sql/walk.go in pkg/sql.walkPlan at line 76
pkg/sql/plan.go in pkg/sql.startExec at line 499
pkg/sql/plan_node_to_row_source.go in pkg/sql.(*planNodeToRowSource).Start at line 124
pkg/sql/distsqlrun/processors.go in pkg/sql/distsqlrun.(*ProcessorBase).Run at line 800
pkg/sql/distsqlrun/flow.go in pkg/sql/distsqlrun.(*Flow).Run at line 626
pkg/sql/distsql_running.go in pkg/sql.(*DistSQLPlanner).Run at line 252
pkg/sql/distsql_running.go in pkg/sql.(*DistSQLPlanner).PlanAndRun at line 839
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).execWithDistSQLEngine at line 1112
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).dispatchToExecutionEngine at line 948
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).execStmtInOpenState at line 456
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).execStmt at line 102
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).run at line 1241
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn at line 433
pkg/sql/pgwire/conn.go in pkg/sql/pgwire.(*conn).serveImpl.func4 at line 337
```

| Tag | Value |
|---|---|
| Cockroach Release | `v19.1.0` |
| Cockroach SHA: | 25dd36f0139bf65b80758deeeccf35ee17ebd622 |
| Platform | linux amd64 |
| Distribution | CCL |
| Environment | v19.1.0 |
| Command | server |
| Go Version | `go1.11.6`|
| # of CPUs |32 |
| # of Goroutines |1602 |

Ok, here's a reproduction case:

```
# Welcome to the CockroachDB demo database!
#
# You are connected to a temporary, in-memory CockroachDB
# instance. Your changes will not be saved!
#
# Web UI: http://127.0.0.1:65133
#
# Server version: CockroachDB CCL v19.2.0-alpha.20190606-500-gad623c864e-dirty (x86_64-apple-darwin18.5.0, built 2019/07/02 16:46:55, go1.12.4) (same version as client)
# Cluster ID: 70d59bbd-57aa-407c-8755-e4efe7f60780
#
# Enter \? for a brief introduction.
#
root@127.0.0.1:65132/defaultdb> create table a (a int primary key, b int);
CREATE TABLE

Time: 2.48ms

root@127.0.0.1:65132/defaultdb> insert into a values(1,1);
INSERT 1

Time: 3.644ms

root@127.0.0.1:65132/defaultdb> begin; alter table a add column c int not null default 5;
Now adding input for a multi-line SQL transaction client-side (smart_prompt enabled).
Press Enter two times to send the SQL text collected so far to the server, or Ctrl+C to cancel.
You can also use \show to display the statements entered so far.
                             ->
ALTER TABLE

Time: 2.832ms

root@127.0.0.1:65132/defaultdb  OPEN> upsert into a select * from a where a=1;
E190702 23:14:09.177349 850 sql/conn_executor.go:713  [n1,client=127.0.0.1:65134,user=root] a SQL panic has occurred while executing "UPSERT INTO a SELECT * FROM a WHERE a = 1": runtime error: index out of range
E190702 23:14:09.177431 850 util/log/crash_reporting.go:207  [n1,client=127.0.0.1:65134,user=root] a panic has occurred!
panic: runtime error: index out of range [recovered]
        panic: panic while executing 1 statements: UPSERT INTO _ SELECT * FROM _ WHERE _ = _; caused by runtime error: index out of range

goroutine 850 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00080bb00, 0x767d280, 0xc001132880, 0x69e8ae0, 0x96c6a00)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:727 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc00080bb00, 0x767d280, 0xc001132880)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:440 +0x61
panic(0x69e8ae0, 0x96c6a00)
        /usr/local/Cellar/go/1.12.4/libexec/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).processKV(0xc000685a00, 0x767d340, 0xc0009db9e0, 0xc001762168, 0x4, 0x15, 0xc001762176, 0x7, 0x7, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:795 +0x1234
github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).NextRow(0xc000685a00, 0x767d340, 0xc0009db9e0, 0x0, 0x10000, 0xc0014536f8, 0x56b7337, 0xc00012f470, 0x1, 0xc00176216a)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:1089 +0xe4
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*rowFetcherWrapper).Next(0xc00034f900, 0x1, 0xc001762168, 0x3, 0x15)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/stats.go:139 +0x4a
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*joinReader).performLookup(0xc0006ee000, 0x2, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/joinreader.go:407 +0xeb
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*joinReader).Next(0xc0006ee000, 0xc000a99980, 0x0, 0x0, 0xc000a99980)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/joinreader.go:297 +0x2b8
github.com/cockroachdb/cockroach/pkg/sql.(*rowSourceToPlanNode).Next(0xc001100900, 0x767d340, 0xc0009db9e0, 0xc00155a200, 0xc00080be50, 0x76ede60, 0x6c16920, 0xc001453a58)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/row_source_to_plan_node.go:75 +0x48
github.com/cockroachdb/cockroach/pkg/sql.(*upsertNode).BatchedNext(0xc0012e02a0, 0x767d340, 0xc0009db9e0, 0xc00155a200, 0xc00080be50, 0x698ee00, 0xc0003e90b0, 0xc001453ae8)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/upsert.go:307 +0xb6
github.com/cockroachdb/cockroach/pkg/sql.(*rowCountNode).startExec(0xc00034f7e0, 0x767d340, 0xc0009db9e0, 0xc00155a200, 0xc00080be50, 0x1, 0xc001453b30)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_batch.go:169 +0xcd
```

--- a/pkg/sql/distsql_physical_planner.go
+++ b/pkg/sql/distsql_physical_planner.go
@@ -1912,8 +1912,9 @@ func (dsp *DistSQLPlanner) createPlanForLookupJoin(
 	}
 
 	joinReaderSpec := distsqlpb.JoinReaderSpec{
-		Table: *n.table.desc.TableDesc(),
-		Type:  n.joinType,
+		Table:      *n.table.desc.TableDesc(),
+		Type:       n.joinType,
+		Visibility: n.table.colCfg.visibility.toDistSQLScanVisibility(),
 	}
 	joinReaderSpec.IndexIdx, err = getIndexIdx(n.table)
 	if err != nil {

--- a/pkg/sql/distsqlrun/joinreader.go
+++ b/pkg/sql/distsqlrun/joinreader.go
@@ -125,10 +125,6 @@ func newJoinReader(
 	post *distsqlpb.PostProcessSpec,
 	output RowReceiver,
 ) (*joinReader, error) {
-	if spec.Visibility != distsqlpb.ScanVisibility_PUBLIC {
-		return nil, errors.AssertionFailedf("joinReader specified with visibility %+v", spec.Visibility)
-	}
-
 	jr := &joinReader{
 		desc:                 spec.Table,
 		input:                input,
@@ -144,13 +140,14 @@ func newJoinReader(
 	if err != nil {
 		return nil, err
 	}
-	jr.colIdxMap = jr.desc.ColumnIdxMap()
+	returnMutations := spec.Visibility == distsqlpb.ScanVisibility_PUBLIC_AND_NOT_PUBLIC
+	jr.colIdxMap = jr.desc.ColumnIdxMapWithMutations(returnMutations)
 
 	var columnIDs []sqlbase.ColumnID
 	columnIDs, jr.indexDirs = jr.index.FullColumnIDs()
 	indexCols := make([]uint32, len(columnIDs))
 	jr.indexTypes = make([]types.T, len(columnIDs))
-	columnTypes := jr.desc.ColumnTypesWithMutations(true)
+	columnTypes := jr.desc.ColumnTypesWithMutations(returnMutations)
 	for i, columnID := range columnIDs {
 		indexCols[i] = uint32(columnID)
 		jr.indexTypes[i] = columnTypes[jr.colIdxMap[columnID]]
@@ -188,15 +185,15 @@ func newJoinReader(
 		collectingStats = true
 	}
 
-	if isSecondary && !jr.neededRightCols().SubsetOf(getIndexColSet(jr.index, jr.colIdxMap)) {
+	neededRightCols := jr.neededRightCols()
+	if isSecondary && !neededRightCols.SubsetOf(getIndexColSet(jr.index, jr.colIdxMap)) {
 		return nil, errors.Errorf("joinreader index does not cover all columns")
 	}
 
 	var fetcher row.Fetcher
 	_, _, err = initRowFetcher(
 		&fetcher, &jr.desc, int(spec.IndexIdx), jr.colIdxMap, false, /* reverse */
-		jr.neededRightCols(), false /* isCheck */, &jr.alloc,
-		distsqlpb.ScanVisibility_PUBLIC,
+		neededRightCols, false /* isCheck */, &jr.alloc, spec.Visibility,
 	)
 	if err != nil {
 		return nil, err

--- a/pkg/sql/logictest/testdata/logic_test/upsert
+++ b/pkg/sql/logictest/testdata/logic_test/upsert
@@ -959,3 +959,30 @@ DO UPDATE SET c = table35970.a+1
 RETURNING b
 ----
 NULL
+
+# ------------------------------------------------------------------------------
+# Regression for #38627: make sure that UPSERTs in the presence of column
+# mutations don't cause problems.
+# ------------------------------------------------------------------------------
+
+statement ok
+CREATE TABLE table38627 (a INT PRIMARY KEY, b INT); INSERT INTO table38627 VALUES(1,1)
+
+statement ok
+BEGIN; ALTER TABLE table38627 ADD COLUMN c INT NOT NULL DEFAULT 5
+
+statement ok
+UPSERT INTO table38627 SELECT * FROM table38627 WHERE a=1
+
+query II
+SELECT * from table38627
+----
+1  1
+
+statement ok
+COMMIT
+
+query III
+SELECT * from table38627
+----
+1  1  5

--- a/pkg/sql/opt/exec/execbuilder/testdata/upsert
+++ b/pkg/sql/opt/exec/execbuilder/testdata/upsert
@@ -378,3 +378,42 @@ INSERT INTO t35364 (x) VALUES (1.5) ON CONFLICT (x) DO UPDATE SET x=2.5 RETURNIN
 
 statement error pq: failed to satisfy CHECK constraint \(z >= 7\)
 UPSERT INTO t35364 (x) VALUES (0)
+
+# ------------------------------------------------------------------------------
+# Regression for #38627. Combined with the equivalent logic test, make sure that
+# UPSERT in the presence of column mutations uses a lookup join without a
+# problem.
+# ------------------------------------------------------------------------------
+
+statement ok
+CREATE TABLE table38627 (a INT PRIMARY KEY, b INT); INSERT INTO table38627 VALUES(1,1)
+
+statement ok
+BEGIN; ALTER TABLE table38627 ADD COLUMN c INT NOT NULL DEFAULT 5
+
+query TTTTT
+EXPLAIN (VERBOSE) UPSERT INTO table38627 SELECT * FROM table38627 WHERE a=1
+----
+count                       ·                      ·                   ()                     ·
+ └── upsert                 ·                      ·                   ()                     ·
+      │                     into                   table38627(a, b)    ·                      ·
+      │                     strategy               opt upserter        ·                      ·
+      └── render            ·                      ·                   (a, b, a, b, c, b, a)  ·
+           │                render 0               a                   ·                      ·
+           │                render 1               b                   ·                      ·
+           │                render 2               a                   ·                      ·
+           │                render 3               b                   ·                      ·
+           │                render 4               c                   ·                      ·
+           │                render 5               b                   ·                      ·
+           │                render 6               a                   ·                      ·
+           └── lookup-join  ·                      ·                   (a, b, a, b, c)        ·
+                │           table                  table38627@primary  ·                      ·
+                │           type                   inner               ·                      ·
+                │           equality               (a) = (a)           ·                      ·
+                │           equality cols are key  ·                   ·                      ·
+                └── scan    ·                      ·                   (a, b)                 ·
+·                           table                  table38627@primary  ·                      ·
+·                           spans                  /1-/1/#             ·                      ·
+
+statement ok
+COMMIT

